File: src/create/objectpool/Pool.java
package create.objectpool;

import java.util.HashSet;
import java.util.Set;

/**
 * Created by Sun on 4/2/2017.
 */

/**
 * Generic object pool
 * @param <T> Type T of Object in the Pool
 */
public abstract class Pool<T> {

    private Set<T> available = new HashSet<>();
    private Set<T> inUse = new HashSet<>();

    protected abstract T create();

    /**
     * Checkout object from pool
     */
    public synchronized T checkOut() {
        if (available.size() <= 0) {
            available.add(create());
        }
        T instance = available.iterator().next();
        available.remove(instance);
        inUse.add(instance);
        return instance;
    }

    public synchronized void checkIn(T instance) {
        inUse.remove(instance);
        available.add(instance);
    }

    @Override
    public String toString() {
        return String.format("Pool available=%d inUse=%d", available.size(), inUse.size());
    }
}

File: src/create/objectpool/Pool.kt
package create.objectpool

import java.util.HashSet

/**
 * Created by Sun on 4/2/2017.
 */

/**
 * Generic object pool
 * @param <T> Type T of Object in the Pool
</T> */
abstract class Pool<T> {

    private val available = HashSet<T>()
    private val inUse = HashSet<T>()

    protected abstract fun create(): T

    /**
     * Checkout object from pool
     */
    @Synchronized fun checkOut(): T {
        if (available.size <= 0) {
            available.add(create())
        }
        val instance = available.iterator().next()
        available.remove(instance)
        inUse.add(instance)
        return instance
    }

    @Synchronized fun checkIn(instance: T) {
        inUse.remove(instance)
        available.add(instance)
    }

    override fun toString(): String {
        return String.format("Pool available=%d inUse=%d", available.size, inUse.size)
    }
}
